From cacdfbefaf5f0d8c570425c369dae1764e9afa51 Mon Sep 17 00:00:00 2001
From: Tomaz Zaman <tomaz@mono.si>
Date: Sat, 18 Apr 2026 23:44:03 +0200
Subject: [PATCH 1/2] plat/nxp/ls1046a/gateway-dk: add board support

Adds platform definitions, build integration and boot glue for the
Mono Gateway Development Kit (LS1046A-based). Supports both QSPI and
eMMC boot paths. DDR initialization is added in a follow-up patch.

Upstream-Status: Inappropriate [product-specific]

Signed-off-by: Tomaz Zaman <tomaz@mono.si>
---
 plat/nxp/soc-ls1046a/gateway_dk/plat_def.h    | 79 +++++++++++++++++++
 plat/nxp/soc-ls1046a/gateway_dk/platform.c    | 28 +++++++
 plat/nxp/soc-ls1046a/gateway_dk/platform.mk   | 36 +++++++++
 .../nxp/soc-ls1046a/gateway_dk/platform_def.h | 13 +++
 plat/nxp/soc-ls1046a/gateway_dk/policy.h      | 16 ++++
 5 files changed, 172 insertions(+)
 create mode 100644 plat/nxp/soc-ls1046a/gateway_dk/plat_def.h
 create mode 100644 plat/nxp/soc-ls1046a/gateway_dk/platform.c
 create mode 100644 plat/nxp/soc-ls1046a/gateway_dk/platform.mk
 create mode 100644 plat/nxp/soc-ls1046a/gateway_dk/platform_def.h
 create mode 100644 plat/nxp/soc-ls1046a/gateway_dk/policy.h

diff --git a/plat/nxp/soc-ls1046a/gateway_dk/plat_def.h b/plat/nxp/soc-ls1046a/gateway_dk/plat_def.h
new file mode 100644
index 000000000..5a0c20b56
--- /dev/null
+++ b/plat/nxp/soc-ls1046a/gateway_dk/plat_def.h
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2025 Mono Technologies Inc.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef PLAT_DEF_H
+#define PLAT_DEF_H
+
+#include <arch.h>
+/*
+ * Required without TBBR.
+ * To include the defines for DDR PHY Images.
+ */
+#include <tbbr_img_def.h>
+
+#include "policy.h"
+#include <soc.h>
+
+#define NXP_SPD_EEPROM0		0x51
+
+#define NXP_SYSCLK_FREQ		100000000
+#define NXP_DDRCLK_FREQ		100000000
+
+/* UART related definition */
+#define NXP_CONSOLE_ADDR	    0x021C0500
+#define NXP_CONSOLE_BAUDRATE	115200
+
+/* Size of cacheable stacks */
+#if defined(IMAGE_BL2)
+#if defined(TRUSTED_BOARD_BOOT)
+#define PLATFORM_STACK_SIZE	0x2000
+#else
+#define PLATFORM_STACK_SIZE	0x1000
+#endif
+#elif defined(IMAGE_BL31)
+#define PLATFORM_STACK_SIZE	0x1000
+#endif
+
+/* SD block buffer */
+#define NXP_SD_BLOCK_BUF_SIZE	(0x8000)
+#define NXP_SD_BLOCK_BUF_ADDR	ULL(0x80000000)
+
+#define BL2_LIMIT		(NXP_OCRAM_ADDR + NXP_OCRAM_SIZE)
+
+/* IO defines as needed by IO driver framework */
+#define MAX_IO_DEVICES		U(3)
+#define MAX_IO_BLOCK_DEVICES	U(1)
+#define MAX_IO_HANDLES		U(4)
+
+/*
+ * FIP image defines - Offset at which FIP Image would be present
+ * Image would include Bl31 , Bl33 and Bl32 (optional)
+ */
+#ifdef POLICY_FUSE_PROVISION
+#define MAX_FIP_DEVICES		U(2)
+#endif
+
+#ifndef MAX_FIP_DEVICES
+#define MAX_FIP_DEVICES		U(1)
+#endif
+
+/*
+ * ID of the secure physical generic timer interrupt used by the BL32.
+ */
+#define BL32_IRQ_SEC_PHY_TIMER	29
+
+/*
+ * Define properties of Group 1 Secure and Group 0 interrupts as per GICv3
+ * terminology. On a GICv2 system or mode, the lists will be merged and treated
+ * as Group 0 interrupts.
+ */
+#define PLAT_LS_G1S_IRQ_PROPS(grp) \
+    INTR_PROP_DESC(BL32_IRQ_SEC_PHY_TIMER, GIC_HIGHEST_SEC_PRIORITY, grp, \
+            GIC_INTR_CFG_LEVEL)
+
+#define PLAT_LS_G0_IRQ_PROPS(grp)
+
+#endif /* PLAT_DEF_H */
diff --git a/plat/nxp/soc-ls1046a/gateway_dk/platform.c b/plat/nxp/soc-ls1046a/gateway_dk/platform.c
new file mode 100644
index 000000000..c99517d77
--- /dev/null
+++ b/plat/nxp/soc-ls1046a/gateway_dk/platform.c
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2025 Mono Technologies Inc.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <plat_common.h>
+
+#pragma weak board_enable_povdd
+#pragma weak board_disable_povdd
+
+bool board_enable_povdd(void)
+{
+#ifdef CONFIG_POVDD_ENABLE
+    return true;
+#else
+    return false;
+#endif
+}
+
+bool board_disable_povdd(void)
+{
+#ifdef CONFIG_POVDD_ENABLE
+    return true;
+#else
+    return false;
+#endif
+}
diff --git a/plat/nxp/soc-ls1046a/gateway_dk/platform.mk b/plat/nxp/soc-ls1046a/gateway_dk/platform.mk
new file mode 100644
index 000000000..419a5b3d9
--- /dev/null
+++ b/plat/nxp/soc-ls1046a/gateway_dk/platform.mk
@@ -0,0 +1,36 @@
+#
+# Copyright 2025 Mono Technologies Inc.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+# Board-specific build parameters
+BOOT_MODE			?=	qspi
+BOARD				:=	gateway_dk
+POVDD_ENABLE		:=	no
+SUPPORTED_BOOT_MODE	:=	qspi emmc
+
+# DDR Compilation Configs
+CONFIG_STATIC_DDR	:= 1
+CONFIG_DDR_NODIMM	:= 1
+DDRC_NUM_DIMM		:= 0
+NUM_OF_DDRC			:= 1
+DDRC_NUM_CS			:= 1
+DDR_ECC_EN			:= yes
+DDR_DEBUG			:= no
+
+# On-Board Flash Details
+QSPI_FLASH_SZ		:= 0x20000000
+NOR_FLASH_SZ		:= 0x20000000
+
+# Platform specific features.
+WARM_BOOT			:= no
+
+# Adding Platform files build files
+BL2_SOURCES			+= ${BOARD_PATH}/ddr_init.c ${BOARD_PATH}/platform.c
+
+# Adding platform board build info
+include plat/nxp/common/plat_make_helper/plat_common_def.mk
+
+# Adding SoC build info
+include plat/nxp/soc-ls1046a/soc.mk
diff --git a/plat/nxp/soc-ls1046a/gateway_dk/platform_def.h b/plat/nxp/soc-ls1046a/gateway_dk/platform_def.h
new file mode 100644
index 000000000..443f34b82
--- /dev/null
+++ b/plat/nxp/soc-ls1046a/gateway_dk/platform_def.h
@@ -0,0 +1,13 @@
+/*
+ * Copyright 2025 Mono Technologies Inc.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef PLATFORM_DEF_H
+#define PLATFORM_DEF_H
+
+#include <plat_def.h>
+#include <plat_default_def.h>
+
+#endif /* PLATFORM_DEF_H */
diff --git a/plat/nxp/soc-ls1046a/gateway_dk/policy.h b/plat/nxp/soc-ls1046a/gateway_dk/policy.h
new file mode 100644
index 000000000..c6c325fa8
--- /dev/null
+++ b/plat/nxp/soc-ls1046a/gateway_dk/policy.h
@@ -0,0 +1,16 @@
+/*
+ * Copyright 2022 NXP
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#ifndef POLICY_H
+#define	POLICY_H
+
+/* Set this to 0x0 to leave the default SMMU page size in sACR
+ * Set this to 0x1 to change the SMMU page size to 64K
+ */
+#define POLICY_SMMU_PAGESZ_64K 0x1
+
+#endif /* POLICY_H */
-- 
2.47.3

